### PR TITLE
Fix endpoint slice filtering to ensure we get all the necessary objects

### PR DESCRIPTION
**Pull request by odinuge**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

Fixes: #issue-number

```release-note
Fix endpoint slices filtering to ensure we filter out headless services and continue to support older k8s versions where service labels are not propagated to endpoint slices
```

### CODE REVIEW

File: pkg/k8s/watchers/watcher.go
Comment (odinuge) [inline pkg/k8s/watchers/watcher.go]: Not sure if we should plumb it in here or just just run `GetEndpointSliceListOptionsModifier` where we need it.

Comment (joamaki) [reply pkg/k8s/watchers/watcher.go]: Fine here. The PR https://github.com/cilium/cilium/pull/23977 removes this from here anyway. I'll rebase and incorporate these changes once this PR lands.

**Review by joamaki — APPROVED** [review]

**Comment by odinuge** [comment]
Hmm. Seems like https://github.com/cilium/cilium/pull/23977 is going along, so I'll let this sit for now. Do you mind pinging when things are stable so I can rebase this PR @joamaki? 😄

**Comment by ti-mo** [comment]
@odinuge thank you for your contribution! I think this would be useful to land in 1.14, but we can hold off until after the release has been cut and treat this as a backport. The tree is currently considered frozen, so now would be a good time to rebase. :smile:

**Comment by ldelossa** [comment]
@odinuge, this PR is suffering from two tests that are stuck. This can be fixed by rebasing this PR again `main` and pushing back up. Can you do this please?

**Comment by squeed** [comment]
@odinuge do you have time to rebase this?

I realize all the box-checking is annoying, and we really appreciate the contribution. I'm sorry it takes so much clicking to get stuff in :disappointed:.

**Comment by odinuge** [comment]
Yes, I'll rebase when I'm back at my computer at work later today. I've been OOTO for a few weeks (it was also in my gh status until yesterday), so I have not been able to rebase during that time.

**Comment by odinuge** [comment]
Rebased now, tho. some earlier changes changed fully how this works, so this won't be possible to backport now 😞 

It does however looks like https://github.com/cilium/cilium/pull/23977 incorrectly changed how this works, adding this filtering with the wrong arguments... I'll debug some more, but that feels like a big release blocker if its the case..

**Comment by odinuge** [comment]
> It does however looks like https://github.com/cilium/cilium/pull/23977 incorrectly changed how this works, adding this filtering with the wrong arguments... I'll debug some more, but that feels like a big release blocker if its the case..

Yes, so for endpointslices, labels from service to endpointSlice will only be done for k8s v1.20+ (https://github.com/kubernetes/kubernetes/pull/94443). However, for endpoints, it has been done for _significantly_ longer (https://github.com/kubernetes/kubernetes/blame/v1.27.0/pkg/controller/endpoint/endpoints_controller.go#L461-L479).

And as I understand it, cilium still wants to support k8s versions older than v1.20, right? (given I read the current code right)

And to be clear; this PR should mitigate this, and k8s pre v1.20 should start working again. Then, we can concider adding the proxy-name filtering to endpointslices when its safe. Should probably add some more docs and a TODO about that.

Any thought on path forwards @joamaki @squeed?

**Comment by odinuge** [comment]
On that note, i'd be curious about the expected behavior of cilium with `--k8s-service-proxy-name` set, and `.toService` egress rules are in used (https://github.com/cilium/cilium/pull/13036/files), given it will change what `services` and `endpoints{,lices}` we get from the apiserver... But thats up for another day.

**Comment by odinuge** [comment]
ping @joamaki ^

**Comment by squeed** [comment]
@odinuge when you mean "k8s pre v1.20 should start working again", you mean that "k8s pre v1.20 with a custom service proxy", right? As in, we no longer rely on that particular label to be propagated from Service to EndpointSlice?

**Comment by odinuge** [comment]
> @odinuge when you mean "k8s pre v1.20 should start working again", you mean that "k8s pre v1.20 with a custom service proxy", right? As in, we no longer rely on that particular label to be propagated from Service to EndpointSlice?

Yes

**Comment by odinuge** [comment]
Mostly wondering if you prefer we merge this PR to get the "intended" behavior for k8s v1.19 and older using a non-nil service proxy value, or if you just want to restore the old behavior with something like this;? 

```diff
diff --git a/pkg/k8s/resource_ctors.go b/pkg/k8s/resource_ctors.go
index ab9dda727e..34bab4ebf5 100644
--- a/pkg/k8s/resource_ctors.go
+++ b/pkg/k8s/resource_ctors.go
@@ -208,8 +208,8 @@ func (lw *endpointsListerWatcher) getListerWatcher() cache.ListerWatcher {
                                lw.cs.Slim().CoreV1().Endpoints(""),
                        )
                        lw.sourceObj = &slim_corev1.Endpoints{}
+                       lw.cachedListerWatcher = utils.ListerWatcherWithModifier(lw.cachedListerWatcher, lw.optsModifier)
                }
-               lw.cachedListerWatcher = utils.ListerWatcherWithModifier(lw.cachedListerWatcher, lw.optsModifier)
        })
        return lw.cachedListerWatcher
 }
 ```

**Comment by squeed** [comment]
@odinuge no, the PR looks good as-is. Just making sure I understood the conditions. Thanks!

**Comment by squeed** [comment]
@odinuge can you please add the Signed-off-by footer in your commit message? Then we can get this in.

**Comment by squeed** [comment]
Likewise, it seems like the PR description is no longer accurate; by my reading of the code, we also dropped headless endpointslices before. The bug fix now is to make service name work on <v1.20. Right?

**Comment by odinuge** [comment]
> @odinuge can you please add the Signed-off-by footer in your commit message? Then we can get this in.

I pushed a new commit yesterday with `Signed-off-by`, so that comment is pointing to an old commit 😄 

> Likewise, it seems like the PR description is no longer accurate; by my reading of the code, we also dropped headless endpointslices before. The bug fix now is to make service name work on <v1.20. Right?

Yee. I could also update the commit messages (I did add an `edit` to [this one](https://github.com/cilium/cilium/pull/25351/commits/385f75cd3edba3fb8d763bf95a55c0eaeda46f83)) again as well, but defer to you.

For the changelog I think its fine/useful having stating the new filtering, given the previous PR that changed the filtering was only a refactor that I assume wasn't intentionally changing the filtering.

**Comment by squeed** [comment]
> Oooh, or did `v1.14` get its own branch already, and `main` is now whats getting into `v1.15`?

bingo. but this should be an easy "backport".

**Comment by squeed** [comment]
> Yee. I could also update the commit messages (I did add an `edit` to [this one](https://github.com/cilium/cilium/pull/25351/commits/385f75cd3edba3fb8d763bf95a55c0eaeda46f83)) again as well, but defer to you.

Please do update the commit messages (and maybe squash it to one?). I realize this PR has been a bit of a saga, but the git way is to forget the journey :-).

We'll get this in soon. I appreciate the fix.

**Comment by squeed** [comment]
I would mention that this fixes the specific use case of using services with `service.kubernetes.io/service-proxy-name` set on clusters pre v1.20 -- otherwise it's hard to piece together the context.

**Review by christarazi — APPROVED** [review]

**Review by squeed — APPROVED** [review]

**Comment by odinuge** [comment]
> I would mention that this fixes the specific use case of using services with service.kubernetes.io/service-proxy-name set on clusters pre v1.20 -- otherwise it's hard to piece together the context.

Missed this before the approvals. Still want me to add this to changelog and/or commit messages? Feels like things are pretty overloaded with info already tbh., but happy to add if you still want

**Pull request merged**
